File: Algoritmos_de_programacion_2023/Python/algoritmo#5/src/ejercicio.py
"""
# 5
ÁREA DE UN POLÍGONO
/*

* Crea una única función (importante que sólo sea una) que sea capaz
* de calcular y retornar el área de un polígono.
* - La función recibirá por parámetro sólo UN polígono a la vez.
* - Los polígonos soportados serán Triángulo, Cuadrado y Rectángulo.
* - Imprime el cálculo del área de un polígono de cada tipo.
 */
"""

import os

"""
Área de un triángulo:
La fórmula general es Área = (base * altura) / 2.

Área de un cuadrado:
La fórmula es Área = lado * lado.

Área de un rectángulo:
La fórmula es Área = longitud * ancho.
"""


def borrar() -> None:
    os.system(command="cls" if os.name == "nt" else "clear")


def main(
    poligono: str,
    base: float = 0,
    altura: float = 0,
    lado: float = 0,
    longitud: float = 0,
    ancho: float = 0,
) -> str:
    match poligono:
        case "1":
            # triangulo
            area: float = base * altura / 2
            return "El area del triangulo es: {variable} metros".format(
                variable=float(area)
            )

        case "2":
            # cuadrado
            area: float = lado * lado
            return "El area del cuadrado es: {variable} metros".format(
                variable=float(area)
            )

        case "3":
            # rectangulo
            area: float = longitud * ancho
            return "El area del rectangulo es: {variable} metros".format(
                variable=float(area)
            )


if __name__ == "__main__":
    while True:
        borrar()
        print(
            """
1. Triangulo
2. Cuadrado
3. Rectangulo

Selecciona el poligono que deseas calcular su area (1, 2, 3)

""",
            end="",
        )
        input_usuario: str = str(input("Seleccione el nombre del poligono: ")).lower()
        match input_usuario:
            case "1":
                while True:
                    try:
                        base_float: float = float(
                            input("Ingrese la base del triangulo: ")
                        )
                        altura_float: float = float(
                            input("Ingrese la altura del triangulo: ")
                        )
                        break
                    except ValueError:
                        print("Ingrese un numero valido")
                        continue
                print(main(poligono=input_usuario, base=base_float, altura=10))

            case "2":
                while True:
                    try:
                        lado_float: float = float(
                            input("Ingrese el lado del cuadrado: ")
                        )
                        break
                    except ValueError:
                        print("Ingrese un numero valido")
                        continue
                print(main(poligono=input_usuario, lado=lado_float))
            case "3":
                while True:
                    try:
                        longitud_float: float = float(
                            input("Ingrese la longitud del rectangulo: ")
                        )
                        ancho_float: float = float(
                            input("Ingrese el ancho del rectangulo: ")
                        )
                        break
                    except ValueError:
                        print("Ingrese un numero valido")
                        continue
                print(
                    main(
                        poligono=input_usuario,
                        longitud=longitud_float,
                        ancho=ancho_float,
                    )
                )
            case _:
                continue
        break

File: Algoritmos_de_programacion_2023/Java/algoritmo_Numero5/src/Ejercicio.java
package Algoritmos_de_programacion_2023.Java.algoritmo_Numero5.src;

import java.util.Scanner;

/*
# 5
ÁREA DE UN POLÍGONO

* Crea una única función (importante que sólo sea una) que sea capaz
* de calcular y retornar el área de un polígono.
* - La función recibirá por parámetro sólo UN polígono a la vez.
* - Los polígonos soportados serán Triángulo, Cuadrado y Rectángulo.
* - Imprime el cálculo del área de un polígono de cada tipo.
*/

/*

*    Área de un triángulo:
*    La fórmula general es Área = (base * altura) / 2.
*
*    Área de un cuadrado:
*    La fórmula es Área = lado * lado.
*
*    Área de un rectángulo:
*    La fórmula es Área = longitud * ancho.

*/
public class Ejercicio {

    public static void borrar() {
        System.out.println("Ingrese un numero valido");
        System.out.print("\033[H\033[2J");
        System.out.flush();
        /*
         * Este código imprime el carácter de escape \033[H que mueve el cursor al
         * inicio de la pantalla y el carácter de escape \033[2J que limpia la pantalla
         * desde el cursor hasta el final. Luego, System.out.flush() asegura que los
         * caracteres de escape se envíen a la terminal inmediatamente.
         */

        /*
         * \033[H: Este es un código de escape CSI (Control Sequence Introducer). El [H
         * es un comando que mueve el cursor a la posición de inicio (esquina superior
         * izquierda de la terminal).
         * 
         * \033[2J: Este es otro código de escape CSI. El [2J es un comando que borra la
         * pantalla desde el cursor hasta el final de la pantalla.
         * 
         * Link de informacion(https://en.wikipedia.org/wiki/ANSI_escape_code)
         */

        return;
    }

    public static String funcion() {
        Scanner scanner = new Scanner(System.in);
        Double base, altura, lado, longitud, ancho;
        int opcion;

        while (true) {
            while (true) {
                try {
                    System.out.println("Que poligono desea calcular su area?");
                    System.out.println("1. Triangulo");
                    System.out.println("2. Cuadrado");
                    System.out.println("3. Rectangulo");
                    opcion = Integer.parseInt(System.console().readLine());
                    if (opcion >= 1) {
                        break;
                    } else {
                        continue;
                    }

                } catch (NumberFormatException e) {
                    System.out.println("Ingrese una opcion valida");
                    scanner.nextLine();
                    borrar();
                    continue;
                }
            }
            switch (opcion) {
                case 1:
                    while (true) {
                        try {
                            System.out.println("Ingrese la base del triangulo");
                            base = Double.parseDouble(System.console().readLine());
                            if (base >= 1.0) {
                                break;
                            } else {
                                continue;
                            }

                        } catch (NumberFormatException e) {
                            System.out.println("Ingrese un numero valido");
                            scanner.nextLine();
                            borrar();
                            continue;
                        }
                    }
                    while (true) {
                        try {
                            System.out.println("Ingrese la altura del triangulo");
                            altura = Double.parseDouble(System.console().readLine());
                            if (altura >= 1.0) {
                                break;
                            } else {
                                continue;
                            }
                        } catch (NumberFormatException e) {
                            System.out.println("Ingrese un numero valido");
                            scanner.nextLine();
                            borrar();
                            continue;
                        }
                    }

                    scanner.close();
                    return "La area del triangulo es " + String.valueOf((base * altura) / 2);

                case 2:
                    while (true) {
                        try {
                            System.out.println("Ingrese el lado del cuadrado");
                            lado = Double.parseDouble(System.console().readLine());
                            if (lado >= 1) {
                                break;
                            } else {
                                continue;
                            }

                        } catch (NumberFormatException e) {
                            System.out.println("Ingrese un numero valido");
                            scanner.nextLine();
                            borrar();
                            continue;
                        }
                    }
                    scanner.close();
                    return "La area del cuadrado es " + String.valueOf((lado * lado));
                case 3:
                    while (true) {
                        try {
                            System.out.println("Ingrese la longitud del rectangulo");
                            longitud = Double.parseDouble(System.console().readLine());
                            if (longitud >= 1.0) {
                                break;
                            } else {
                                continue;
                            }

                        } catch (NumberFormatException e) {
                            System.out.println("Ingrese un numero valido");
                            scanner.nextLine();
                            borrar();
                            continue;
                        }
                    }
                    while (true) {
                        try {
                            System.out.println("Ingrese la ancho del rectangulo");
                            ancho = Double.parseDouble(System.console().readLine());
                            if (ancho >= 1.0) {
                                break;
                            } else {
                                continue;
                            }
                        } catch (NumberFormatException e) {
                            System.out.println("Ingrese un numero valido");
                            scanner.nextLine();
                            borrar();
                            continue;
                        }
                    }
                    scanner.close();
                    return "La area del rectangulo es " + String.valueOf((longitud * ancho));
                default:
                    break;

            }

        }
        /*
         * Puedes tener varios bloques catch para manejar diferentes tipos de
         * excepciones. También puedes tener un bloque catch que maneje todas las
         * excepciones
         * try {
         * Código que puede lanzar una excepción
         * } catch (ArithmeticException e) {
         * Manejar excepción ArithmeticException
         * } catch (NullPointerException e) {
         * Manejar excepción NullPointerException
         * } catch (Exception e) {
         * Manejar todas las demás excepciones
         * }
         */

        // Para que sirve Integer.parseInt(System.console().readLine())

    }

    public static void main(String[] args) {
        borrar();
        String resultado = funcion();
        System.out.println(resultado);
    }
}

/*
 * 
 * boolean
 * Clase Envolvente: Boolean
 * Valor Predeterminado (Primitivo): false
 * Valor Predeterminado (Objeto): null
 * 
 * 
 * byte
 * Clase Envolvente: Byte
 * Valor Predeterminado (Primitivo): 0
 * Valor Predeterminado (Objeto): null
 * 
 * 
 * short
 * Clase Envolvente: Short
 * Valor Predeterminado (Primitivo): 0
 * Valor Predeterminado (Objeto): null
 * 
 * 
 * int
 * Clase Envolvente: Integer
 * Valor Predeterminado (Primitivo): 0
 * Valor Predeterminado (Objeto): null
 * 
 * 
 * long
 * Clase Envolvente: Long
 * Valor Predeterminado (Primitivo): 0L
 * Valor Predeterminado (Objeto): null
 * 
 *
 * float
 * Clase Envolvente: Float
 * Valor Predeterminado (Primitivo): 0.0f
 * Valor Predeterminado (Objeto): null
 * 
 * 
 * double
 * Clase Envolvente: Double
 * Valor Predeterminado (Primitivo): 0.0d
 * Valor Predeterminado (Objeto): null
 * 
 *
 * char
 * Clase Envolvente: Character
 * Valor Predeterminado (Primitivo): \u0000
 * Valor Predeterminado (Objeto): null
 * 
 * 
 * String
 * Clase: String
 * Valor Predeterminado: null
 */